### PR TITLE
Fix issue #108: Add output length management for script-based tools

### LINKED ISSUE #108: [Performance] Add output length management for script-based tools
# [Performance] Add output length management for script-based tools

## Problem/Motivation

Script-based tools can generate extremely large outputs (stdout/stderr) that can overwhelm the response system, consume excessive memory, or make results difficult to read. Currently, there's no mechanism to limit or manage the length of outputs returned by script-based tools.

**Issues with Large Outputs:**
- **Memory Usage**: Very large outputs consume significant memory
- **Network Transfer**: Large responses slow down API calls
- **UI Performance**: Large text blocks can freeze or slow down user interfaces
- **Readability**: Excessive output makes it hard to find relevant information
- **Storage**: Large outputs may exceed database field limits or storage quotas

## Proposed Solution

Add output length management capabilities to the `post_processing` configuration section for script-based tools in YAML.

### YAML Configuration Structure

```yaml
tools:
  deploy:
    enabled: true
    name: deploy
    type: script
    scripts:
      linux: bash {pwd}/deploy.sh
      darwin: bash {pwd}/deploy.sh
      windows: powershell -file {pwd}/deploy.ps1
    post_processing:
      output_limits:
        max_stdout_length: 5000        # Max characters for stdout
        max_stderr_length: 2000        # Max characters for stderr  
        max_total_length: 6000         # Max combined length
        truncate_strategy: "end"       # "start", "end", "middle", "smart"
        truncate_message: "\\n... (output truncated due to length)"
        preserve_last_lines: 10        # Keep last N lines when truncating
        preserve_first_lines: 5        # Keep first N lines when truncating
        preserve_raw: false            # Keep original output in separate field
    inputSchema:
      type: object
      properties:
        environment:
          type: string
          enum: ["dev", "staging", "prod"]
      required: ["environment"]
```

### Configuration Options

| Option | Type | Default | Description |
|--------|------|---------|-------------|
| `max_stdout_length` | integer | unlimited | Maximum characters for stdout |
| `max_stderr_length` | integer | unlimited | Maximum characters for stderr |
| `max_total_length` | integer | unlimited | Maximum combined length of both outputs |
| `truncate_strategy` | string | `"end"` | Truncation strategy: `"start"`, `"end"`, `"middle"`, `"smart"` |
| `truncate_message` | string | `"\\n... (truncated)"` | Message to insert where content was truncated |
| `preserve_last_lines` | integer | 0 | Number of lines to preserve at the end |
| `preserve_first_lines` | integer | 0 | Number of lines to preserve at the beginning |
| `preserve_raw` | boolean | false | Keep original untruncated output in `raw_output`/`raw_error` fields |

#### `"end"` (default)
Keep beginning, truncate end:
```
Original: "Line 1\\nLine 2\\nLine 3\\n...very long output...\\nLine 1000"
Result:   "Line 1\\nLine 2\\nLine 3\\n... (truncated)"
```

#### `"start"`
Keep end, truncate beginning:
```
Original: "Line 1\\nLine 2\\nLine 3\\n...very long output...\\nLine 1000"
Result:   "... (truncated)\\nLine 998\\nLine 999\\nLine 1000"
```

#### `"middle"`
Keep beginning and end, truncate middle:
```
Original: "Line 1\\nLine 2\\nLine 3\\n...very long output...\\nLine 1000"  
Result:   "Line 1\\nLine 2\\n... (truncated)\\nLine 999\\nLine 1000"
```

#### `"smart"`
Intelligent truncation preserving important sections:
- Preserve error messages and warnings
- Keep stack traces intact
- Maintain log structure
- Prioritize recent output

## Acceptance Criteria

- [ ] Script-based tools can be configured with `post_processing.output_limits`
- [ ] `max_stdout_length` limits stdout to specified character count
- [ ] `max_stderr_length` limits stderr to specified character count
- [ ] `max_total_length` limits combined output to specified character count
- [ ] All four truncation strategies work correctly (`end`, `start`, `middle`, `smart`)
- [ ] `preserve_last_lines` and `preserve_first_lines` work with line-based preservation
- [ ] `truncate_message` is inserted at truncation points
- [ ] `preserve_raw: true` keeps original output in separate fields
- [ ] Truncation only occurs when outputs exceed specified limits
- [ ] All existing script-based tools continue to work unchanged (backward compatibility)
- [ ] Output limiting is completely optional

### Files to Modify

1. **`mcp_tools/yaml_tools.py`**
   - Extend `YamlToolBase._execute_script()` method
   - Add `_apply_output_limits()` method
   - Integrate limiting into post-processing pipeline

2. **`mcp_tools/output_limiter.py`** (new file)
   - Implement different truncation strategies
   - Handle line-based preservation
   - Provide smart truncation logic

3. **`server/tools.yaml`**
   - Add examples with output limiting configurations

### Implementation Approach

```python
def _apply_output_limits(self, result: Dict, limits: Dict) -> Dict:
    \"\"\"Apply output length limits to command result.\"\"\"
    if not limits:
        return result
    
    stdout = result.get(\"output\", \"\")
    stderr = result.get(\"error\", \"\")
    
    # Preserve raw outputs if requested
    if limits.get(\"preserve_raw\", False):
        result[\"raw_output\"] = stdout
        result[\"raw_error\"] = stderr
    
    # Apply individual limits
    max_stdout = limits.get(\"max_stdout_length\")
    max_stderr = limits.get(\"max_stderr_length\")
    
    if max_stdout and len(stdout) > max_stdout:
        result[\"output\"] = self._truncate_text(stdout, max_stdout, limits)
    
    if max_stderr and len(stderr) > max_stderr:
        result[\"error\"] = self._truncate_text(stderr, max_stderr, limits)
    
    # Apply total limit
    max_total = limits.get(\"max_total_length\")
    if max_total:
        current_total = len(result.get(\"output\", \"\")) + len(result.get(\"error\", \"\"))
        if current_total > max_total:
            result[\"output\"], result[\"error\"] = self._truncate_combined(
                result.get(\"output\", \"\"), result.get(\"error\", \"\"), max_total, limits
            )
    
    return result

def _truncate_text(self, text: str, max_length: int, limits: Dict) -> str:
    \"\"\"Truncate text using specified strategy.\"\"\"
    if len(text) <= max_length:
        return text
    
    strategy = limits.get(\"truncate_strategy\", \"end\")
    message = limits.get(\"truncate_message\", \"\\n... (truncated)\")
    
    if strategy == \"end\":
        return self._truncate_end(text, max_length, message, limits)
    elif strategy == \"start\":
        return self._truncate_start(text, max_length, message, limits)
    elif strategy == \"middle\":
        return self._truncate_middle(text, max_length, message, limits)
    elif strategy == \"smart\":
        return self._truncate_smart(text, max_length, message, limits)
    
    return text[:max_length - len(message)] + message
```

### 1. Build Tool (preserve recent output)
```yaml
build_project:
  type: script
  scripts:
    linux: make build
  post_processing:
    output_limits:
      max_stdout_length: 4000
      truncate_strategy: \"start\"
      preserve_last_lines: 20
      truncate_message: \"\\n... (earlier build output truncated)\\n\"
```

### 2. Log Analysis Tool (preserve both ends)
```yaml
analyze_logs:
  type: script
  scripts:
    linux: python {pwd}/analyze.py
  post_processing:
    output_limits:
      max_total_length: 5000
      truncate_strategy: \"middle\"
      preserve_first_lines: 5
      preserve_last_lines: 10
      truncate_message: \"\\n... (middle section truncated) ...\\n\"
```

### 3. Deployment Tool (smart truncation with raw preservation)
```yaml
deploy:
  type: script
  scripts:
    linux: bash {pwd}/deploy.sh
  post_processing:
    output_limits:
      max_stdout_length: 3000
      max_stderr_length: 1000
      truncate_strategy: \"smart\"
      preserve_raw: true
      truncate_message: \"\\n... (output truncated for readability) ...\\n\"
```

### 4. Test Runner (separate limits for stdout/stderr)
```yaml
run_tests:
  type: script
  scripts:
    linux: pytest --verbose
  post_processing:
    output_limits:
      max_stdout_length: 2000  # Test results summary
      max_stderr_length: 5000  # Detailed error output
      truncate_strategy: \"end\"
      preserve_last_lines: 15
```

## Performance Considerations

1. **Memory Efficiency**: Process outputs in chunks when possible
2. **Line-Based Processing**: Work with lines rather than characters for better preservation
3. **Lazy Evaluation**: Only apply limits when outputs exceed thresholds
4. **Caching**: Cache truncation results for repeated access
5. **Regex Optimization**: Use efficient patterns for smart truncation

## Smart Truncation Logic

The `\"smart\"` strategy should:
- Preserve error messages and stack traces
- Keep log timestamps and structure
- Prioritize recent output over old output
- Maintain context around important events
- Preserve JSON/XML structure boundaries

```python
def _truncate_smart(self, text: str, max_length: int, message: str, limits: Dict) -> str:
    \"\"\"Apply intelligent truncation preserving important content.\"\"\"
    lines = text.split('\\n')
    
    # Identify important lines (errors, warnings, etc.)
    important_patterns = [
        r'\\b(error|exception|failed|failure)\\b',
        r'\\b(warning|warn)\\b',
        r'\\b(stack trace|traceback)\\b',
        r'^\\d{4}-\\d{2}-\\d{2}.*\\b(error|exception|failed)\\b'  # timestamped errors
    ]
    
    # Implementation details for smart preservation...
    # This would be a more complex algorithm
```

## Additional Context

This enhancement addresses performance and usability issues with large script outputs while providing flexible control over truncation behavior. The system is designed to be:

- **Performance conscious**: Efficient processing of large outputs
- **User-friendly**: Configurable truncation strategies for different use cases
- **Information preserving**: Smart truncation maintains important content
- **Backward compatible**: Completely optional with sensible defaults

**Related Files:**
- `mcp_tools/yaml_tools.py` (lines 211-426: `_execute_script` method)
- `server/tools.yaml` (script-based tool examples)
- `mcp_tools/command_executor/executor.py` (output handling)

### PR DESCRIPTION
**Pull request by plusplusoneplusplus**
## Summary

This PR implements output length management for script-based tools as requested in issue #108.

**Developed using Cursor**

### Core Implementation
- **New `OutputLimiter` class** (`mcp_tools/output_limiter.py`):
  - Multiple truncation strategies: `end`, `start`, `middle`, `smart`
  - Support for `max_stdout_length`, `max_stderr_length`, `max_total_length` limits
  - Line-based preservation options (`preserve_first_lines`, `preserve_last_lines`)
  - Smart truncation that preserves error messages and warnings
  - Raw output preservation option (`preserve_raw`)

### Integration
- **Enhanced `yaml_tools.py`**:
  - Integrated output limiting into post-processing pipeline
  - Applied after security filtering but before output attachment config
  - Backward compatible - completely optional

### Documentation & Examples
- **Comprehensive example** (`mcp_tools/tool.example.yaml`):
  - Single tool demonstrating all output limiting capabilities
  - Matches actual `tools.yaml` structure and format
  - Shows integration with existing post-processing options
  - Includes usage examples for different scenarios

### Configuration Structure
```yaml
post_processing:
  output_limits:
    max_stdout_length: 5000        # Max characters for stdout
    max_stderr_length: 2000        # Max characters for stderr  
    max_total_length: 6000         # Max combined length
    truncate_strategy: "smart"     # "start", "end", "middle", "smart"
    truncate_message: "\n... (output truncated due to length)"
    preserve_last_lines: 10        # Keep last N lines when truncating
    preserve_first_lines: 5        # Keep first N lines when truncating
    preserve_raw: false            # Keep original output in separate field
```

### Truncation Strategies
1. **"end"** (default): Keep beginning, truncate end
2. **"start"**: Keep end, truncate beginning  
3. **"middle"**: Keep beginning and end, truncate middle
4. **"smart"**: Intelligent truncation preserving important content (errors, warnings, stack traces)

### Testing
- **Comprehensive test suite** with 23 test cases covering:
  - All truncation strategies
  - Line preservation options
  - Combined limits (stdout + stderr)
  - Edge cases and error handling
  - Raw output preservation
  - All tests passing ✅

## Performance Considerations
- **Memory Efficient**: Only applies limits when outputs exceed thresholds
- **Line-Based Processing**: Works with lines rather than characters for better preservation
- **Lazy Evaluation**: Processes outputs in chunks when possible

## Backward Compatibility
- ✅ All existing script-based tools continue to work unchanged
- ✅ Output limiting is completely optional
- ✅ No breaking changes to existing configurations

## Files Added/Modified
- `mcp_tools/output_limiter.py` - Core OutputLimiter class
- `mcp_tools/yaml_tools.py` - Integration with post-processing pipeline
- `mcp_tools/tests/test_output_limiter.py` - Comprehensive test suite
- `mcp_tools/tool.example.yaml` - Complete configuration example

## Fixes
Closes #108

## Testing Instructions
1. Copy the example from `mcp_tools/tool.example.yaml` into your `tools.yaml`
2. Configure a script tool with `output_limits` in `post_processing`
3. Run the script and verify output is truncated according to strategy
4. Test different truncation strategies and line preservation options
5. Verify raw output preservation when enabled

### CODE REVIEW

File: mcp_tools/tool.example.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      # NEW: Output length management - all options demonstrated
+      output_limits:
+        # Length limits (all optional, no limits if not specified)
+        max_stdout_length: 5000        # Maximum characters for stdout
+        max_stderr_length: 2000        # Maximum characters for stderr
+        max_total_length: 6000         # Maximum combined length (applied after individual limits)
+
+        # Truncation strategy (default: "end")
+        # Available options:
+        #   "end" - Keep beginning, truncate end (good for logs, build output)
```
Comment (plusplusoneplusplus) [inline mcp_tools/tool.example.yaml:64]: can you make the default start (preserve the end)?

**Comment by plusplusoneplusplus** [comment]
✅ **Addressed PR Comment**

I've updated the default truncation strategy from `"end"` to `"start"` as requested. This change:

**Changes Made:**
- **OutputLimiter default**: Changed from `"end"` to `"start"` strategy
- **Example documentation**: Updated `tool.example.yaml` to reflect new default
- **Usage examples**: Reordered to show `"start"` as the default behavior

**Why "start" is better as default:**
- Preserves **final results** and **completion status**
- Shows **error messages** that typically appear at the end
- Keeps **exit codes** and **summary information**
- More useful for most command-line tools and scripts

**Backward Compatibility:**
- ✅ All existing configurations continue to work unchanged
- ✅ Users can still explicitly specify `truncate_strategy: "end"` if needed
- ✅ All 23 tests still pass

The `"start"` strategy is indeed more practical as the default since most users care about seeing how their commands ended rather than how they started.
